src/chips/sam4l/src/spi.rs

```diff
@@ -419,52 +415,38 @@ impl spi_master::SpiMaster for Spi {
 
 impl DMAClient for Spi {
     fn xfer_done(&mut self, pid: usize) {
-        // I don't know if there are ordering guarantees on the read and
-        // write interrupts, guessing not, so issue the callback when both
-        // reading and writing are complete. In practice it seems like
-        // the read interrupt happens second. -pal
-        //
-        // The disable calls are commented out because executing them
-        // causes subsequent operations to fail. The call to read_write_bytes
-        // calls enable(), so I don't know why. -pal
-        if pid == 4 {
-            // SPI RX
+        // We ignore the RX interrupt because we are guaranteed to have TX
+        // DMA setup, but there's no guarantee RX will exist. In the case
+        // both are happening, just using TX is sufficient because SPI
+        // is full duplex.
+        if pid == 22 {
+            // SPI TX
+            self.writing.set(false);
             self.reading.set(false);
-            let buf = match self.dma_read.as_mut() {
+
+            let txbuf = match self.dma_write.as_mut() {
                 Some(dma) => {
                     let buf = dma.abort_xfer();
                     dma.disable();
                     buf
                 }
                 None => None,
             };
-            self.read_buffer = buf;
-            if !self.reading.get() && !self.writing.get() {
-                let rb = self.read_buffer.take();
-                let wb = self.write_buffer.take();
-                let len = self.dma_length.get();
-                self.dma_length.set(0);
-                self.callback.as_ref().map(|cb| cb.read_write_done(wb, rb, len));
-            }
-        } else if pid == 22 {
-            // SPI TX
-            self.writing.set(false);
-            let buf = match self.dma_write.as_mut() {
+
+            let rxbuf = match self.dma_read.as_mut() {
```

Comment:
I tried to change this, compiler gets upset: `error: cannot move out of borrowed content`